### PR TITLE
Fixed Gobblin Yarn to be able to run MultiWorkUnits

### PR DESCRIPTION
**Pull request by liyinan926**
Originally Gobblin on Yarn was written with the assumption that we would rely on Helix to group/pack `WorkUnit`s into groups and assign them properly to the available containers. So the code originally flattened the `MultiWorkUnit`s. Later it was realized this would not be too practical since Helix does not have the knowledge of loads of individual `WorkUnit`s. In addition, the Gobblin Kafka adapter is already doing bin packing so we decided to not override that logic and changed the code to not flatten `MultiWorkUnit`s. However, the parts that actually run the job/tasks are not updated accordingly to work with `MultiWorkUnit`s. This PR fixed it. 

Signed-off-by: Yinan Li <liyinan926@gmail.com>

### CODE REVIEW

**Comment by liyinan926** [comment]
@zliu41 can you review?

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java [hunk 1]
```diff
@@ -98,82 +93,97 @@ public GobblinHelixTask(TaskCallbackContext taskCallbackContext, TaskExecutor ta
...
-      switch (workingState) {
-        case SUCCESSFUL:
-          return new TaskResult(TaskResult.Status.COMPLETED, "task id: " + this.taskId);
-        case FAILED:
-          return new TaskResult(TaskResult.Status.ERROR, "task id: " + this.taskId);
-        case CANCELLED:
-          return new TaskResult(TaskResult.Status.CANCELED, "task id: " + this.taskId);
-        default:
-          throw new IllegalStateException("Unexpected result task state: " + workingState);
+      Path workUnitFilePath =
```
Comment (sahilTakiar) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: A lot of the code below seems similar to what is done in `MRJobLauncher.TaskRunner` anyway we can abstract the code to a new class so we can re-use it for both MR jobs and YARN jobs?

Comment (liyinan926) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: Sure. Will do.

File: gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java [hunk 2]
```diff
@@ -98,82 +93,97 @@ public GobblinHelixTask(TaskCallbackContext taskCallbackContext, TaskExecutor ta
...
+      runWorkUnits(workUnits);
+
+      return new TaskResult(TaskResult.Status.COMPLETED, String.format("completed tasks: %d", workUnits.size()));
     } catch (IOException ioe) {
-      LOGGER.error("Failed to build task " + this.taskId, ioe);
-      return new TaskResult(TaskResult.Status.ERROR, "task id: " + this.taskId);
-    } catch (ExecutionException ee) {
-      LOGGER.error("Failed to run task " + this.taskId, ee);
-      return new TaskResult(TaskResult.Status.ERROR, "task id: " + this.taskId);
+      return new TaskResult(TaskResult.Status.ERROR, Throwables.getStackTraceAsString(ioe));
```
Comment (sahilTakiar) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: Is it necessary to have a catch for `IOException` as it is handled in the same was as `Throwable`

Comment (liyinan926) [inline gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixTask.java]: Good catch. It used to need this but no longer the case now.

**Comment by liyinan926** [comment]
@sahilTakiar Updated.

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java [hunk 3]
```diff
@@ -520,6 +482,117 @@ private void cleanupStagingData(JobState jobState) {
...
+   * @param jobId the job ID
+   * @param workUnits the given list of {@link WorkUnit}s to submit to run
+   * @param taskStateTracker a {@link TaskStateTracker} for task state tracking
+   * @param taskExecutor a {@link TaskExecutor} for task execution
+   * @param taskStateStore a {@link StateStore} for storing {@link TaskState}s
+   * @param logger a {@link Logger} for logging
+   * @throws IOException if there's something wrong with any IO operations
+   * @throws InterruptedException if the task execution gets cancelled
+   */
+  public static void runWorkUnits(String jobId, String containerId, List<WorkUnit> workUnits,
```
Comment (sahilTakiar) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: Any reason to make this method `static`? It looks like any class that is using this method will extend `AbstractJobLauncher` anyway. So we can just make this `protected void runWorkUnits(...)`.

Same for the method below.

Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: Actually no. `GobblinHelixTask` and `MRJobLauncher.TaskRunner` use it and both of them do not extend `AbstractJobLauncher`.

**Comment by sahilTakiar** [comment]
LGTM

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java [hunk 4]
```diff
@@ -520,6 +482,117 @@ private void cleanupStagingData(JobState jobState) {
...
+      TaskStateTracker taskStateTracker,  TaskExecutor taskExecutor, StateStore<TaskState> taskStateStore,
+      Logger logger) throws IOException, InterruptedException {
+
+    if (workUnits.isEmpty()) {
+      logger.warn("No work units to run in container " + containerId);
```
Comment (zliu41) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: You are using both `LOG` and `logger` in this class.

Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: This is intentional so the logger specific to the classes that call this method can be used.

File: gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java [hunk 5]
```diff
@@ -520,6 +482,117 @@ private void cleanupStagingData(JobState jobState) {
...
+   * </p>
+   *
+   * @param jobId the job ID
+   * @param workUnits the given list of {@link WorkUnit}s to submit to run
+   * @param stateTracker a {@link TaskStateTracker} for task state tracking
+   * @param taskExecutor a {@link TaskExecutor} for task execution
+   * @param countDownLatch a {@link java.util.concurrent.CountDownLatch} waited on for job completion
+   * @return a list of {@link Task}s from the {@link WorkUnit}s
+   */
+  public static List<Task> runWorkUnits(String jobId, List<WorkUnit> workUnits, TaskStateTracker stateTracker,
```
Comment (zliu41) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: This method can be private

Comment (liyinan926) [inline gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java]: This method is called directly in `LocalJobLauncher`.

**Pull request merged**